### ISSUE TITLE
`getChromInfoFromEnsembl` is erroring for GRCh37 genome build

### ISSUE THREAD

**Issue by mjsteinbaugh** [Bioconductor/GenomeInfoDb#97]
Hi Bioconductor team, I noticed an edge case issue with `getChromInfoFromEnsembl`. When attempting to get `Seqinfo` for GRCh37, not specifying a release works as expected, but specifying the last supported release for GRCh37 (87) currently errors.

This works as expected:

```r
GenomeInfoDb::getChromInfoFromEnsembl(
    species = "Homo sapiens",
    use.grch37 = TRUE,
    as.Seqinfo = TRUE
)
```

This is currently broken, due a file parsing issue on the Ensembl FTP server:

```r
GenomeInfoDb::getChromInfoFromEnsembl(
    species = "Homo sapiens",
    release = 87L,
    use.grch37 = TRUE,
    as.Seqinfo = TRUE
)
```

```
Error in scan(file = file, what = what, sep = sep, quote = quote, dec = dec,  : 
  line 498 did not have 10 elements
Calls: <Anonymous> ... simple_read_table -> do.call -> do.call -> <Anonymous> -> scan
Backtrace:
     ▆
  1. └─GenomeInfoDb::getChromInfoFromEnsembl(...)
  2.   └─GenomeInfoDb:::.get_chrom_info_from_Ensembl_FTP(...)
  3.     └─GenomeInfoDb:::fetch_seq_regions_from_Ensembl_FTP(...)
  4.       └─GenomeInfoDb:::.fetch_synonyms_from_Ensembl_FTP(...)
  5.         └─GenomeInfoDb:::fetch_table_from_Ensembl_FTP(core_db_url, "external_db")
  6.           └─GenomeInfoDb:::fetch_table_from_url(...)
  7.             └─GenomeInfoDb:::simple_read_table(destfile, ...)
  8.               ├─BiocGenerics::do.call(read.table, args)
  9.               ├─base::do.call(read.table, args)
 10.               └─utils (local) `<fn>`(...)
 11.                 └─base::scan(...)
```

Should we contact Ensembl to get this fixed on the FTP server?

Best,
Mike

**Comment by hpages** [comment]
Same as issue #98 but this time with the `external_db.txt.gz` file from https://ftp.ensembl.org/pub/grch37/release-87/mysql/homo_sapiens_core_87_37/

Unfortunately this one seems to break `data.table::fread()` so, unfortunately, my previous fix (see commit fe81f0d423f07efabb6e61d6c57620171a6b75b7) doesn't help here. Bummer :disappointed: 

Another way to go would be to query the Ensembl MySQL db directly but given how flaky/slow that service tends to be I'm not very optimistic. For GRCh37, one needs to query the MySQL server running at ensembldb.ensembl.org on port 3337 (as explained here https://useast.ensembl.org/info/data/mysql.html):
```
fetch_external_db_for_grch37_release87 <- function(drv) {
    host <- "ensembldb.ensembl.org"
    port <- 3337
    username <- "anonymous"
    dbname <- "homo_sapiens_core_87_37"
    dbconn <- dbConnect(drv, host=host, port=port, username=username, dbname=dbname)
    on.exit(dbDisconnect(dbconn))
    dbGetQuery(dbconn, "SELECT * FROM external_db")
}
```
Let's see:
```
library(RMariaDB)
extdb87 <- fetch_external_db_for_grch37_release87(RMariaDB::MariaDB())
# Error: Failed to connect: Lost connection to MySQL server at 'waiting for initial communication packet', system error: 110
```
Hmm.. let's try with **RMySQL**:
```
library(RMySQL)
extdb87 <- fetch_external_db_for_grch37_release87(RMySQL::MySQL())
# Warning message:
# In dbSendQuery(conn, statement, ...) :
#   Unsigned INTEGER in col 0 imported as numeric

dim(extdb87)
# [1] 587  10
```
Ok, that _kind of_ works. BUT:
- My understanding is that **RMySQL** has been superseded by **RMariaDB** so it's kind of disappointing that the latter fails.
- I've actually played a little bit with these things before and in my experience direct queries to Ensembl MySQL server tends to not be very reliable. It's also quite slow. See https://github.com/Bioconductor/GenomeInfoDb/blob/fe81f0d423f07efabb6e61d6c57620171a6b75b7/R/fetch_table_from_Ensembl_FTP.R#L103-L113
  So I'm reluctant to drop completely the other methods (`read.table` & `fread`) in `fetch_table_from_Ensembl_FTP()`, and to have the function use the direct db query method for everything.
- This means that the direct db method would be an alternative method implemented in `fetch_table_from_Ensembl_FTP()`. However getting this right is probably not as straightforward as it sounds.

So again...

:thinking: 

... maybe cleaning the tab file before calling `read.table()` is the way to go? I wish I knew how to do this in R.

H.

**Comment by hpages** [comment]
I forgot about `readBin()`. It does the trick here:
```
## Remove \r characters (carriage returns) from a file.
clean_file <- function(infile, outfile)
{
    con1 <- file(infile, "rb")
    on.exit(close(con1))
    con2 <- file(outfile, "wb")
    on.exit(close(con2), add=TRUE)
    while (TRUE) {
        bytes <- readBin(con1, what=raw(), n=1000L)
        if (length(bytes) == 0L)
            break
        bytes <- bytes[bytes != charToRaw("\r")]
        writeBin(bytes, con2)
    }
}
```
I'll integrate this to `GenomeInfoDb:::fetch_table_from_Ensembl_FTP()` soon.

**Comment by hpages** [comment]
Should be fixed in **GeomeInfoDb** 1.38.5 (release) and 1.39.5 (devel). There were a few additional complications. See commit 58da6591fb92a8665a83b755bbab7cbc54c1b965.

Let me know how that works for you.

**Comment by hpages** [comment]
Hopefully everything works fine for you now.
